kvdb/postgres/readwrite_tx.go

```diff
@@ -0,0 +1,392 @@
+// +build kvdb_postgres
+
+package postgres
+
+import (
+	"bytes"
+	"database/sql"
+	"sort"
+	"strings"
+
+	"github.com/btcsuite/btcwallet/walletdb"
+)
+
+// readWriteTx holds a reference to an open postgres transaction.
+type readWriteTx struct {
+	readOnly bool
+	db       *db
+	tx       *sql.Tx
+
+	// onCommit gets called upon commit.
+	onCommit func()
+
+	// active is true if the transaction hasn't been committed yet.
+	active bool
+}
+
+// newReadWriteTx creates an rw transaction using a connection from the
+// specified pool.
+func newReadWriteTx(db *db, readOnly bool) (*readWriteTx, error) {
+	// Obtain the global lock instance. An alternative here is to obtain a
+	// database lock from Postgres. Unfortunately there is no database-level
+	// lock in Postgres, meaning that each table would need to be locked
+	// individually. Perhaps an advisory lock could perform this function
+	// too.
+	if readOnly {
```

Comment:
There's a cool pattern that I learned recently that could be used here. Instead of storing the `readOnly` flag and having the `if tx.readOnly` in several places, we could store a `sync.Locker` on the TX that is either read or write:
```
	var locker sync.Locker = &db.lock
	if readOnly {
		locker = db.lock.RLocker()
	}
	locker.Lock()

	tx, err := db.db.Begin()
	if err != nil {
		locker.Unlock()
		return nil, err
	}

	return &readWriteTx{
		db:       db,
		tx:       tx,
		active:   true,
		locker: locker,
	}, nil
```

Reply:
Nice! Applied.